### ISSUE TITLE
Literals as types 

### ISSUE THREAD

**Issue by aspnetde** [fsharp/fslang-suggestions#1195]
I propose we implement something like what's known as [Literal Types in TypeScript](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types):

```ts
const foo = (bar: "A" | "B") => {}

foo("A"); // ok
foo("C"); // forbidden
```

However, as in TypeScript, it should not be limited to literals. In fact, TS accepts any type here:

```ts
interface Options {
  width: number;
}
function configure(x: Options | "auto") {}
configure({ width: 100 }); // ok
configure("auto"); // ok
configure("automatic"); // forbidden
```

As someone who is actively working on an application for the last couple of months which uses TS for the frontend and F# for the backend, I find myself often missing this capability when switching from TS to F#. Right now, I have to explicitly define a discriminated union to do that in F#:

```fs
type Bar = A | B
let foo (bar: Bar) = ()
foo A
```

Instead, it would be really neat to define that DU "on the fly" or "inline" (not sure what the proper wording might be) as in TypeScript:
```fs
let foo (bar: A | B) = ()
```

## Pros and Cons 

Advantages: More flexibility, I guess.

Disadvantages: As always, it's not a silver bullet. There are plenty of reasons one may want to define the type of that parameter explicitly, e.g., to reuse it elsewhere.

## Affidavit

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by kerams** [comment]
https://github.com/fsharp/fslang-suggestions/issues/538
https://github.com/dotnet/fsharp/pull/12265

**Comment by aspnetde** [comment]
> #538 [dotnet/fsharp#12265](https://github.com/dotnet/fsharp/pull/12265)

Looks like it is too early in the morning and my ability to do a search properly is still limited before the first coffee. [RFC FS-1092](https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1092-anonymous-type-tagged-unions.md) indeed looks exactly like it.

**Comment by dsyme** [comment]
The suggestion here is misnamed, it's actually for typescript style literals-as-types

**Comment by dsyme** [comment]
https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types

**Comment by dsyme** [comment]
There have been previous suggestions along these lines, usually too ungrounded in practice and I rejected them, but we should dig them out. 

The TS experience shows these are highly useful in pragmatic interop scenarios, and I expect Fable would greatly benefit from them.

**Comment by kerams** [comment]
We have that already as well :) https://github.com/fsharp/fslang-suggestions/issues/656

**Comment by dbrattli** [comment]
FYI: Literal types in Python: https://peps.python.org/pep-0586/ . I would think they work mostly the same way as in TS, but syntax is a bit different. You can basically choose if you write `Literal["a", "b"]` or `Literal["a"] | Literal["b"]`. They have the exact same meaning. They can also be composed, e.g:

```py
ReadOnlyMode         = Literal["r", "r+"]
WriteAndTruncateMode = Literal["w", "w+", "wt", "w+t"]
WriteNoTruncateMode  = Literal["r+", "r+t"]
AppendMode           = Literal["a", "a+", "at", "a+t"]

AllModes = Literal[ReadOnlyMode, WriteAndTruncateMode,
                   WriteNoTruncateMode, AppendMode]
```

**Comment by mindlid** [comment]
just for the record, Scala also [has support for them](https://docs.scala-lang.org/sips/42.type.html)

```scala
// the following constant can only store ints from 1 to 3
val three: 1 | 2 | 3 = 3

val one: 1 = 1                     // val declaration
def foo(x: 1): Option[1] = Some(x) // param type, type arg
def bar[T <: 1](t: T): T = t       // type parameter bound
foo(1: 1)                          // type ascription
```

**Comment by mindlid** [comment]
This may be required for expressing [Nullable Reference Types Suggestion](https://github.com/fsharp/fslang-design/blob/0ff87d5ab3160366a96316f739ba31d9a61d0f4a/RFCs/FS-1060-nullable-reference-types.md?plain=1#L1) in F# as literal Union `string | null`, instead of `string?` the same way [Scala](https://docs.scala-lang.org/scala3/reference/experimental/explicit-nulls.html) and Typescript do
```scala
// Scala
val x: String = null // error: found `Null`, but required `String`
val x: String | Null = null // ok
```
```typescript
// typescript
const x: string = null // Type 'null' is not assignable to type 'string'.
const y: string | null = null //ok
```
both Scala and Typescript have literal types, so Nullable types as unions emerge naturally, as opposed to C# or Kotlin where unions are not available yet

**Comment by njlr** [comment]
I have landed here from https://github.com/fsharp/fslang-suggestions/issues/608

Does this suggestion cover using integer literals in types?

This would allow for encoding dimensions. For example:

```fsharp
let u : Vector<2, float> = v2 3.0 4.0
let v : Vector<3, float> = v3 1.0 2.0 3.0

let w = u + v // Compile-time error
```

**Comment by Happypig375** [comment]
.NET 9 introduces [const generics](https://github.com/dotnet/runtime/issues/89730) which is highly relevant here. C# will also include a design about this. We can refer to them for implementing this feature.

**Comment by vzarytovskii** [comment]
> .NET 9 introduces [const generics](https://github.com/dotnet/runtime/issues/89730) which is highly relevant here. C# will also include a design about this. We can refer to them for implementing this feature.

It's not decided yet. Hasn't been discussed, championed, or even reviewed.

I will be surprised if it'll make it to 9 tbh.

Also, if we'll be using it, we'll be constraining runtime version, need to keep it in mind.

**Comment by Happypig375** [comment]
@vzarytovskii The only platform that we'll leave out is effectively only .NET Framework since people upgrade to newer .NET runtimes nowadays.

**Comment by vzarytovskii** [comment]
> @vzarytovskii The only platform that we'll leave out is effectively only .NET Framework since people upgrade to newer .NET runtimes nowadays.

Not in my experience. Recently was working with one of biggest F# customers, which *just* migrated to net6.

**Comment by Happypig375** [comment]
@vzarytovskii By the time this is implemented, .NET 10 would have been in LTS for years already.

**Comment by vzarytovskii** [comment]
> @vzarytovskii By the time this is implemented, .NET 10 would have been in LTS for years already.

I was actually thinking taking this and anonymous DUs for F#9

**Comment by Happypig375** [comment]
@vzarytovskii
If you have influence over runtime and C# design then probably you can ensure that polyfills are allowed across C# and F#. F# can then implement with polyfills.

**Comment by vzarytovskii** [comment]
> @vzarytovskii
> If you have influence over runtime and C# design then probably you can ensure that polyfills are allowed across C# and F#. F# can then implement with polyfills.

I can probably join ldm when they're going to be discussing it, but probably worth asking question in the runtime suggestion

**Comment by LeaveNhA** [comment]
Any development on this, fellas?

**Comment by vzarytovskii** [comment]
> Any development on this, fellas?

No, it lacks approval, design and RFC, so not really actionable.

**Comment by voronoipotato** [comment]
Kinda glad, literal types feel really cool until you start dealing with two different DU with the same literal types. It's a little extra typing but that typing provides disambiguation, and it does not take long to have two literal DU with the same cases.  In fact I see an example in this thread.

**Comment by SchlenkR** [comment]
> The TS experience shows these are highly useful in pragmatic interop scenarios [...]

What other advantages would literal types in F# bring? I say: (as good as) none. Only through the combination of set-like types and their structural character, narrowing and a bottom type do literal types unfold power. In TypeScript, all these features exist, which then allows DUs and exhaustiveness checks to be encoded. In F#, we already have DUs, and: Without having the mentioned features - apart from other trivial things - I can't think of any other useful examples. Are there any?

**Comment by atsapura** [comment]
Working around this is very easy: just declare DU in 1 line and if, let's say, you'd like to have a numeric literal as type, you can just create a `this.ToInt()` method. What's the point of polluting the language?

**Comment by Lanayx** [comment]
This will greatly help custom DSL for views, for example Oxpecker.ViewEngine:
```fsharp
input(type'="text") // compiles fine
input(type'="abcd") // compilation error, since it will be literal union violation
```

**Comment by voronoipotato** [comment]
you can use typeproviders to provide additional type check errors if you want.

**Comment by voronoipotato** [comment]
This would be better addressed by the smart constructor feature.

**Comment by voronoipotato** [comment]
It would cause way more harm than it could help and this can be achieved with other methods, smart constructors, myriad code generation, or even SRTP

**Comment by Lanayx** [comment]
> It would cause way more harm than it could help and this can be achieved with other methods, smart constructors, myriad code generation, or even type providers

In my case it can't be achieved by any of the tools you mentioned

**Comment by Lanayx** [comment]
> ```fsharp
> type A = B | C;;
> 
> > nameof C;;
> val it: string = "C"
> ```
> 
> same as your literals as types for strings but with less typing and lower risk

I think `nameof` doesn't apply here, it is not type-safe, it is more typing, or if you meant the API side, it just can't be used there. The closest is using DU in API and using `.ToString` on the API side, but it's more typing (and overall DX is worse due to many different DUs with same cases), less readable, worse performance-wise and doesn't allow custom extensibility (since with literal type user might want to ignore warning and pass arbitrary string)

**Comment by NatElkins** [comment]
Just want to chime in that I would love this for TS/JS interop with Fable.

**Comment by voronoipotato** [comment]
@Lanayx if you want it to be type safe don't use the literal string until you're ready to use it. The proposal is a nonstarter due to how it would introduce inconsistency into the type system. We don't want to make the type system unsound in the pursuit of convenience or comfort. I can see several new kinds of type error that would no longer be possible to be caught by the compiler in the examples of this thread alone. I understand that introduces more typing, but this proposal would create vastly more work than it reduces in it's current form. If you can find a way where a type literal of "A" from one DU and a type literal of "A" from an entirely different DU aren't ambiguous I might be more interested.

**Comment by Lanayx** [comment]
I agree that this is a nonstarter until [anonymous union types](https://github.com/fsharp/fslang-suggestions/issues/538) are implemented first, so not being in rush. And when implemented, the string literal type logic should be quite close to handling `null` literal.

**Comment by vzarytovskii** [comment]
> the string literal type logic should be quite close to handling null literal

Well, yes and no, but largely no. Null is both property of existing subset of types (reference types) and an existing constraint (at least from the F# type system perspective). 

Literal types would be whole new F# only constraint. Which will not be fitting neither nulls nor anonymous unions.

**Comment by voronoipotato** [comment]
This suggestion does hint to me that there is an unmet need to prevent boiler plate on the interpreting side. This is something I do occasionally run up against and could alleviate half of the problem. I might make a proper suggestion for it later on. 

```fsharp
type Bar = A | B 
let comparer (s: string) : int =  String.Compare(str, arg, StringComparison.InvariantCultureIgnoreCase)

let (a: Bar option) = Bar.tryParse "A"  //tryParse would be autogenerated by the compiler
let (b: Bar option) = Bar.tryParseWith comparer "b"

//alternatively
let (a': Bar option) = Union.tryParse<Bar> "A"
let (b': Bar option) = Union.tryParseWith<Bar> comparer "b"
```

**Comment by voronoipotato** [comment]
> Just want to chime in that I would love this for TS/JS interop with Fable.

@NatElkins This change could be done in Fable without any change to the F# language, and I wouldn't argue against it (as much) because TS interop makes good sense for Fable. I would probably still want a setting to fail the build if it's used because literals as types can be an extremely bad footgun that would probably take out half of my coworkers.
